client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala

```diff
@@ -75,6 +77,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
   val latestPartitionLocation =
     JavaUtils.newConcurrentHashMap[Int, ConcurrentHashMap[Int, PartitionLocation]]()
   private val userIdentifier: UserIdentifier = IdentityProvider.instantiate(conf).provide()
+  private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, ArrayBuffer[(Int, Boolean)]]()
```

Comment:
Why not use:
```suggestion
  private val shuffleIdMapping = JavaUtils.newConcurrentHashMap[Int, scala.collection.mutable.LinkedHashMap[Int, Boolean]]()
```

This will avoid all the O(n) `find` scans through this - we can use `lastOption`

Reply:
by design, the max stage retry  number should be small, typically 3 - 4
Given a small size array, I suppose the O(N) lookup/update time should not be a performance issue, and I've seen too much HashMap/ConcurrentHashMap in LifecycleManager ...
I can accept the change if you insist

btw, lastOption is a good idea, an empty collection is unexpected, but it is always good to print a meaningful error message to users
will update

Reply:
I was ofcourse assuming this (bounded size) is not the case - see my comments ([1](https://github.com/apache/incubator-celeborn/pull/1924#discussion_r1375372656), [2](https://github.com/apache/incubator-celeborn/pull/1924#discussion_r1375371317)).
'Typical' case should be small ofcourse !

Reply:
this issue had been handled with new solution as described in https://github.com/apache/incubator-celeborn/pull/1924#discussion_r1381725261